cli/dispatch_minimal.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+// Do not add flatbuffer dependencies to this module.
+//! Connects to js/dispatch_minimal.ts sendAsyncMinimal This acts as a faster
+//! alternative to flatbuffers using a very simple list of int32s to lay out
+//! messages. The first i32 is used to determine if a message a flatbuffer
+//! message or a "minimal" message, see `has_minimal_token()`.
+use crate::state::ThreadSafeState;
+use deno::Buf;
+use deno::Op;
+use deno::PinnedBuf;
+use futures::Future;
+
+const DISPATCH_MINIMAL_TOKEN: i32 = 0xCAFE;
+const OP_READ: i32 = 1;
+const OP_WRITE: i32 = 2;
+
+pub fn has_minimal_token(s: &[i32]) -> bool {
+  s[0] == DISPATCH_MINIMAL_TOKEN
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+// This corresponds to RecordMinimal on the TS side.
+struct Record {
+  pub promise_id: i32,
+  pub op_id: i32,
+  pub arg: i32,
+  pub result: i32,
+}
+
+impl Into<Buf> for Record {
+  fn into(self) -> Buf {
+    let vec = vec![
+      DISPATCH_MINIMAL_TOKEN,
+      self.promise_id,
+      self.op_id,
+      self.arg,
+      self.result,
+    ];
+    let buf32 = vec.into_boxed_slice();
+    let ptr = Box::into_raw(buf32) as *mut [u8; 5 * 4];
+    unsafe { Box::from_raw(ptr) }
+  }
+}
+
+impl From<&[i32]> for Record {
+  fn from(s: &[i32]) -> Record {
+    let ptr = s.as_ptr();
+    let ints = unsafe { std::slice::from_raw_parts(ptr, 5) };
+    assert_eq!(ints[0], DISPATCH_MINIMAL_TOKEN);
+    Record {
+      promise_id: ints[1],
+      op_id: ints[2],
+      arg: ints[3],
+      result: ints[4],
+    }
+  }
+}
+
+pub fn dispatch_minimal(
+  state: &ThreadSafeState,
+  control32: &[i32],
+  zero_copy: Option<PinnedBuf>,
+) -> Op {
+  let record = Record::from(control32);
+  let is_sync = record.promise_id == 0;
+  let min_op = match record.op_id {
+    OP_READ => ops::read(record.arg, zero_copy),
+    OP_WRITE => ops::write(record.arg, zero_copy),
+    _ => unimplemented!(),
+  };
+
+  let state = state.clone();
+  let mut record_ = record;
```

Comment:
If you just declare `record` as mut above then you don't need this one.

Reply:
Done